File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let res = []

    function dfs(i, slate, sum){
        if(sum === target){
            res.push([...slate])
            return
        }
        if(i >= candidates.length || sum > target){
            return
        }

        slate.push(candidates[i])
        dfs(i, slate, sum+candidates[i])
        slate.pop()
        dfs(i+1, slate, sum)
    }

    dfs(0, [], 0)
    return res
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()

        def helper(start, slate, total):
            if total == target:
                res.append(slate[:])
                return
            
            for i in range(start, len(candidates)):
                if total + candidates[i] > target:
                    break
                slate.append(candidates[i])
                helper(i, slate, total + candidates[i])
                slate.pop()
        
        helper(0, [], 0)
        return res
